java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -434,35 +442,60 @@ private <T extends IncrementalTask> void runAndUpdateTask(
       // we want to obscure that has been retried??
       // but this is how FW historically worked, so leaving it
       taskState.clearRetryCount();
-
       taskState.setMostRecentUpdateTime(Instant.now());
-      postRunUpdate = getDatastore().newTransaction();
     } catch (ShardFailureException ex ) {
-      postRunUpdate = getDatastore().newTransaction();
-      retryState = handleShardFailure(postRunUpdate, jobState, taskState, ex);
+      retryType = RetryType.SHARD;
+      t = ex;
     } catch (JobFailureException ex) {
-      postRunUpdate = getDatastore().newTransaction();
-      log.log(Level.WARNING,
-        "Shard " + taskState.getTaskId() + " triggered job failure", ex);
-      handleJobFailure(postRunUpdate, taskState, ex);
+      retryType = RetryType.JOB;
+      t = ex;
+      log.log(Level.WARNING, "Shard " + taskState.getTaskId() + " triggered job failure", ex);
     } catch (RuntimeException ex) {
-      postRunUpdate = getDatastore().newTransaction();
-      retryState = handleSliceFailure(postRunUpdate, jobState, taskState, ex, false);
+      t = ex;
+      retryType = RetryType.SLICE;
     } catch (Throwable ex) {
-      postRunUpdate = getDatastore().newTransaction();
+      t = ex;
+      retryType = RetryType.SLICE; // this was originally shard, but seems like a mistake
```

Comment:
Note this change:
- message (slice error) and action (shard failure) were inconsistent
- probably makes sense to handle it as slice failure rather than prematurely kill the shard?